airbyte-integrations/connectors/source-square/source_square/source.py

```diff
@@ -0,0 +1,446 @@
+        location_ids = [location["id"] for location in locations_records]
+
+        if location_ids:
+            json_payload["location_ids"] = location_ids
+
+        json_payload["limit"] = self.items_per_page_limit
+        return json_payload
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        json_payload = self.request_body_json(stream_state=stream_state, stream_slice=stream_slice, next_page_token=None)
+        if "location_ids" not in json_payload:
+            self.logger.info("No location records found.")
+            yield from []
+
+        # There is a restriction in the documentation where only 10 locations can be send at one request
+        # https://developer.squareup.com/reference/square/orders-api/search-orders#request__property-location_ids
+        location_ids = json_payload["location_ids"]
+        separated_locations = separate_items_by_count(location_ids, 10)
+
+        stream_state = stream_state or {}
+        pagination_complete = False
+
+        for locations in separated_locations:
+            json_payload["location_ids"] = locations
+            next_page_token = None
+            while not pagination_complete:
+                request_headers = self.request_headers(
+                    stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
+                )
+                request = self._create_prepared_request(
+                    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
+                    headers=dict(request_headers, **self.authenticator.get_auth_header()),
+                    params=self.request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
+                    json=json_payload,
+                )
+
+                response = self._send_request(request)
+                yield from self.parse_response(response, stream_state=stream_state, stream_slice=stream_slice)
+
+                next_page_token = self.next_page_token(response)
+                if not next_page_token:
+                    pagination_complete = True
+
+            # Always return an empty generator just in case no records were ever yielded
+            yield from []
+
+
+class SourceSquare(AbstractSource):
+    api_version = "2021-06-16"  # Latest Stable Release
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+
+        headers = {
+            "Square-Version": self.api_version,
+            "Authorization": "Bearer {}".format(config["api_key"]),
+            "Content-Type": "application/json",
+        }
+        url = "https://connect.squareup{}.com/v2/catalog/info".format("sandbox" if config["is_sandbox"] else "")
+
+        try:
```

Comment:
lets do the same 'extended' error handling as we do on basic requests

Reply:
Done.